### PR TITLE
percona-server 5.7.20-18

### PR DESCRIPTION
**Pull request by raymondelooff**
Created with `brew bump-formula-pr`.

### CODE REVIEW

**Comment by xenji** [comment]
The jenkins build fails with a 404 for the patch (see https://jenkins.brew.sh/job/Homebrew%20Core%20Pull%20Requests/version=high_sierra/15057/console)

This part is not needed anymore, because the PR has been merged into master in September and the release from Dec., 11th picked it up.
```
  # Fix C++ build failure due to Xcode 9 being very strict
  if DevelopmentTools.clang_build_version >= 900
    patch do
      url "https://github.com/percona/percona-server/pull/1925.patch?full_index=1"
      sha256 "126ed7762ab94a4b2afdaa8a09d35d5e25dfd7cd5452cf51b4db90144e737e6e"
    end
  end```

I've locally disabled it an ran into a compilation problem I cannot yet sort out: 
```cd /tmp/percona-server-20171221-67953-ri1qi8/percona-server-5.7.20-18/extra/protobuf && /usr/local/Cellar/cmake/3.10.1/bin/cmake -E cmake_link_script CMakeFiles/protoclib.dir/link.txt --verbose=1
/usr/bin/ar qc libprotoclib.a  CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/code_generator.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/command_line_interface.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/plugin.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/plugin.pb.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/subprocess.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/zip_writer.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_enum.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_enum_field.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_extension.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_field.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_file.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_generator.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_helpers.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_message.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_message_field.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_primitive_field.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_service.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/cpp/cpp_string_field.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_context.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_enum.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_enum_field.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_extension.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_field.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_file.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_generator.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_generator_factory.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_helpers.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_lazy_message_field.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_message.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_message_field.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_name_resolver.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_primitive_field.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_shared_code_generator.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_service.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_string_field.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/java/java_doc_comment.cc.o CMakeFiles/protoclib.dir/protobuf-2.6.1/src/google/protobuf/compiler/python/python_generator.cc.o
/usr/bin/ranlib libprotoclib.a
[  5%] Built target protoclib
make: *** [all] Error 2```

I'm digging deeper and come back when I have results.

**Comment by xenji** [comment]
Adding this after the lines that disable TokuDB make it work:
```
# MyROCKS does not compile on macOS
args.concat %w[-DWITHOUT_ROCKSDB=1]
```

**Comment by raymondelooff** [comment]
Thanks for the research @xenji!

File: Formula/percona-server.rb [hunk 1]
```diff
@@ -88,10 +80,14 @@ def install
...
     ]
 
-    # TokuDB is broken on MacOsX
+    # TokuDB is broken on macOS
     # https://bugs.launchpad.net/percona-server/+bug/1531446
     args.concat %w[-DWITHOUT_TOKUDB=1]
 
+    # Percona MyRocks does not compile on macOS
+    # https://www.percona.com/doc/percona-server/LATEST/myrocks/install.html
+    args.concat %w[-DWITHOUT_ROCKSDB=1]
```
Comment (ilovezfs) [inline Formula/percona-server.rb]: Has this been reported? We'll need an issue link here.

**Comment by ilovezfs** [comment]
I assume the RocksDB support is working with 5.7.18-16? If so, I'm a bit reluctant to ship this, especially since the `rocksdb` formula itself is building just fine, so this is probably a question of percona being outdated in their merges of upstream rocksdb commits and should be fixable.

**Comment by raymondelooff** [comment]
The Percona MyRocks storage engine was released in version 5.7.19-17 ([release notes](https://www.percona.com/doc/percona-server/LATEST/release-notes/Percona-Server-5.7.19-17.html)). The current formula is based on version 5.7.18-16. The feature is currently experimental and not officially tested for macOS. I can't find an existing issue on their issue tracker.

**Comment by ilovezfs** [comment]
Cool, that makes sense. Can you please create such an issue? They're apparently not testing on macOS at all since this could have been discovered before release and fixed with the upstream rocksdb fixes.

**Comment by xenji** [comment]
macOS is not an officially supported platform anyway, neither is Windows. They concentrate on Linux only. Reporting makes sense, but I won't bet on a fix in the near future ;).

**Comment by ilovezfs** [comment]
The fix is to update their RocksDB code. This isn't their bug but it is fixed already.

**Comment by ilovezfs** [comment]
Specifcally: https://github.com/facebook/rocksdb/blob/master/port/port_posix.h#L25-L52

**Comment by xenji** [comment]
I could try and test the build against the homebrew version of rocksdb. Hopefully there are no other patches applied by them...

**Comment by raymondelooff** [comment]
I've tried to compile it with RocksDB version 5.6.1, but I got the same error. They seem to be using a version of RocksDB that isn't tagged as a release, but it is close to version 5.6.1: https://github.com/facebook/rocksdb/tree/50a969131f69ffab6f7a694b3b897a7235e899fa. 

The `port_posix.h` file is the same and includes the lines you've mentioned: https://github.com/facebook/rocksdb/blob/50a969131f69ffab6f7a694b3b897a7235e899fa/port/port_posix.h. We can try to compile it with a newer version of RocksDB.

**Comment by ilovezfs** [comment]
Someone will need to report this upstream otherwise it won't be fixed. But I've shipped this as-is.

Thanks @raymondelooff!

**Comment by raymondelooff** [comment]
No problem, thanks! I've found that there is no conditional for `Darwin`, so it doesn't set the `OS_MACOSX` definition: https://github.com/percona/percona-server/blob/5.7/storage/rocksdb/CMakeLists.txt#L78-L89. RocksDB itself has the conditional and sets the `OS_MACOSX` definition: https://github.com/facebook/rocksdb/blob/master/CMakeLists.txt#L312-L318.

I've tried to add a conditional to the CMakeLists.txt file of Percona Server and it fixes the problem mentioned earlier. However, it still doesn't compile due to other errors.

```diff
diff --git a/storage/rocksdb/CMakeLists.txt b/storage/rocksdb/CMakeLists.txt
index c281c57b412..47c41471201 100644
--- a/storage/rocksdb/CMakeLists.txt
+++ b/storage/rocksdb/CMakeLists.txt
@@ -85,6 +85,8 @@ IF(UNIX)
     IF(HAVE_FALLOCATE AND HAVE_LINUX_FALLOC_H)
       ADD_DEFINITIONS(-DROCKSDB_FALLOCATE_PRESENT)
     ENDIF()
+  ELSEIF(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
+    ADD_DEFINITIONS(-DOS_MACOSX)
   ENDIF()
 ENDIF()
```

I also noticed that this formula is quite different from the standard MySQL formula. The MySQL build is a lot smaller, because it removes a large directory with tests (373 MB uncompressed). I will open another PR to bring the formula more in-line with the standard MySQL formula.

**Comment by ilovezfs** [comment]
Nice sleuthing @raymondelooff!

>However, it still doesn't compile due to other errors.

Those should be reported too! :)

**Comment by raymondelooff** [comment]
Compilation errors reported: https://bugs.launchpad.net/percona-server/+bug/1741639

**Comment by ilovezfs** [comment]
Thank you, @raymondelooff!

**Pull request closed without merging**